lib/sql.js

```diff
@@ -1508,3 +1508,263 @@ SQLConnector.prototype.getInsertedId = function(model, info) {
 SQLConnector.prototype.executeSQL = function(sql, params, options, callback) {
   throw new Error(g.f('{{executeSQL()}} must be implemented by the connector'));
 };
+
+// Refactored Discovery methods
+
+SQLConnector.prototype.buildQuerySchemas = function(options) {
+  var sql = 'SELECT catalog_name as "catalog",' +
+   ' schema_name as "schema"' +
+   ' FROM information_schema.schemata';
+  return this.paginateSQL(sql, 'schema_name', options);
+};
+
+SQLConnector.prototype.paginateSQL = function(sql, orderBy, options) {
+  throw new Error(g.f('{{paginateSQL}} must be implemented by the connector'));
+};
+/**
+ * Discover database schemas
+ *
+ // * @param {Object} options Options for discovery
+ * @param {Function} [cb] The callback function
+ */
+SQLConnector.prototype.discoverDatabaseSchemas = function(options, cb) {
+  if (!cb && typeof options === 'function') {
+    cb = options;
+    options = {};
+  }
+  options = options || {};
+  var self = this;
+  this.execute(self.buildQuerySchemas(options), cb);
+};
+
+/*!
+ * Build sql for listing tables
+ * @param options {all: for all owners, owner: for a given owner}
+ * @returns {string} The sql statement
+ */
+SQLConnector.prototype.buildQueryTables = function(options) {
+  throw new Error(g.f('{{buildQueryTables}} must be implemented by the connector'));
+};
+
+/*!
+ * Build sql for listing views
+ * @param options {all: for all owners, owner: for a given owner}
+ * @returns {string} The sql statement
+ */
+SQLConnector.prototype.buildQueryViews = function(options) {
+  throw new Error(g.f('{{buildQueryViews}} must be implemented by the connector'));
+};
+
+/**
+ * Discover model definitions
+ *
+ * @param {Object} options Options for discovery
+ * @param {Function} [cb] The callback function
+ */
+SQLConnector.prototype.discoverModelDefinitions = function(options, cb) {
+  if (!cb && typeof options === 'function') {
+    cb = options;
+    options = {};
+  }
+  options = options || {};
+
+  var self = this;
+  var calls = [function(callback) {
+    self.execute(self.buildQueryTables(options), callback);
+  }];
+
+  if (options.views) {
+    calls.push(function(callback) {
+      self.execute(self.buildQueryViews(options), callback);
+    });
+  }
+  async.parallel(calls, function(err, data) {
+    if (err) {
+      cb(err, data);
+    } else {
+      var merged = [];
+      merged = merged.concat(data.shift());
+      if (data.length) {
+        merged = merged.concat(data.shift());
+      }
+      cb(err, merged);
+    }
+  });
+};
+
+SQLConnector.prototype.buildQueryColumns = function(schema, table) {
+  throw new Error(g.f('{{buildQueryColumns}} must be implemented by the connector'));
```

Comment:
We can use https://github.com/strongloop/loopback-connector-mysql/blob/master/lib/discovery.js#L196 as the default implementation.